src/hotspot/share/logging/circularStringBuffer.hpp

```diff
@@ -0,0 +1,212 @@
+#include "runtime/mutex.hpp"
+#include "runtime/os.inline.hpp"
+#include "runtime/semaphore.hpp"
+#include "utilities/globalDefinitions.hpp"
+#include "utilities/resourceHash.hpp"
+
+#include <string.h>
+#ifdef LINUX
+#include <sys/mman.h>
+#endif
+
+// The CircularMapping is a struct that provides
+// an interface for writing and reading bytes in a circular buffer
+// correctly. This indirection is necessary because there are two
+// underlying implementations: Linux, and all others.
+#ifdef LINUX
+// Implements a circular buffer by using the virtual memory mapping facilities of the OS.
+// Specifically, it reserves virtual memory with twice the size of the requested buffer.
+// The latter half of this buffer is then mapped back to the start of the first buffer.
+// This allows for write_bytes and read_bytes to consist of a single memcpy, as the
+// wrap-around is dealt with by the virtual memory system.
+struct CircularMapping {
+  FILE* file;
+  char* buffer;
+  size_t size;
+
+  CircularMapping()
+  : file(nullptr), buffer(nullptr), size(0) {
+  };
+
+  CircularMapping(size_t size);
+  ~CircularMapping() {
+    munmap(buffer, size * 2);
+    fclose(file);
+  }
+
+  void write_bytes(size_t at, const char* bytes, size_t size) {
+    memcpy(&buffer[at], bytes, size);
+  }
+
+  void read_bytes(size_t at, char* out, size_t size) {
+    memcpy(out, &buffer[at], size);
+  }
+};
+#else
+struct CircularMapping {
+  char* buffer;
+  size_t size;
+  CircularMapping()
+  : buffer(nullptr), size(0) {
+  };
+  CircularMapping(size_t size);
+
+  void write_bytes(size_t at, const char* bytes, size_t size) {
+    const size_t part1_size = MIN2(size, this->size - at);
+    const size_t part2_size = size - part1_size;
+
+    ::memcpy(&buffer[at], bytes, part1_size);
+    ::memcpy(buffer, &bytes[part1_size], part2_size);
+  }
+
+  void read_bytes(size_t at, char* out, size_t size) {
+    const size_t part1_size = MIN2(size, this->size - at);
+    const size_t part2_size = size - part1_size;
+
+    ::memcpy(out, &buffer[at], part1_size);
+    ::memcpy(&out[part1_size], buffer, part2_size);
+  }
+
+  ~CircularMapping() {
+    os::release_memory(buffer, size);
+  }
+};
+#endif
+
+class CircularStringBuffer {
+  friend class AsyncLogTest;
+
+public:
+    // account for dropped messages
+  using StatisticsMap = ResourceHashtable<LogFileStreamOutput*, uint32_t, 17, /*table_size*/
+                                        AnyObj::C_HEAP, mtLogging>;
+private:
+  static const LogDecorations& None;
+  const bool _should_stall; // Should a producer stall until a consumer has made room for its message?
+
+  // Need to perform accounting of statistics under a separate lock.
+  StatisticsMap& _stats;
+  PlatformMonitor& _stats_lock;
+
+  // Can't use a Monitor here as we need a low-level API that can be used without Thread::current().
+  PlatformMonitor _read_lock;
+  PlatformMonitor _write_lock;
+  Semaphore _flush_sem;
+
+  struct ReadLocker : public StackObj {
```

Comment:
The ReadLocker/WriteLocker terminology initially made me think you were using ReadWriteLocks, but IIUC these are simply two distinct locks for different mutating operations. I prefer the producer/consumer naming from the description in JBS.

The overall locking protocol is unclear - the JBS description only uses mutexes not monitors.

Reply:
>The overall locking protocol is unclear - the JBS description only uses mutexes not monitors.

The wait/notify capability is used for announcing that more memory is available or that a new message is available. This isn't necessary for the locking protocol in itself, we could also repeatedly take the respective locks and look at the data in order to determine the state of the system. Wait/notify simply allows us to yield the thread in the cases of a full buffer or no messages being available for printing.

Reply:
The wait/notify becomes a little clearer with producer/consumer terminlogy.